analyzers/src/SonarAnalyzer.CFG/Roslyn/CfgAllPathValidator.cs

```diff
@@ -35,18 +35,34 @@ public abstract class CfgAllPathValidator
             this.cfg = cfg;
 
         public bool CheckAllPaths() =>
-            IsBlockOrAllSuccessorsValid(this.cfg.EntryBlock);
+            IsBlockOrAllSuccessorsValid(cfg.EntryBlock);
 
-        private bool IsBlockOrAllSuccessorsValid(BasicBlock block) =>
-            !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+        private bool IsBlockOrAllSuccessorsValid(BasicBlock block)
+        {
+            var isValid = !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+            visitedStatus[block] = !isValid;
+            return isValid;
+        }
 
         private bool AreAllSuccessorsValid(BasicBlock block)
         {
-            this.visited.Add(block);
-
-            return !block.SuccessorBlocks.Contains(cfg.ExitBlock)
-                   && block.SuccessorBlocks.Except(visited).Any()
-                   && block.SuccessorBlocks.Except(visited).All(IsBlockOrAllSuccessorsValid);
+            visitedStatus[block] = false; // protects from loops, don't fail the computation if hits itself
```

Comment:
Both UTs in CfgAllPathValidatorTest` are green even if I change this to `true` 😒 Can you craft one that would fail?